### PR TITLE
simplifying R package for efficiency and robustness

### PR DESCRIPTION
**Pull request by bfgray3**
* avoid `for` loops whenever possible
* prefer `seq_len` and `seq_along` over `1:n` or `1:length(x)` 
* other small changes

### CODE REVIEW

**Comment by thirdwing** [comment]
Is there any specific reason we can't `library(methods)`?

**Comment by bfgray3** [comment]
In general, it is best not to attach packages when loading another package since it modifies the search path for the existing session and this could cause issues if the user is not aware.  Not the end of the world here, however, since `methods` is now automatically attached when R starts.  I was running into issues trying to isolate everywhere I needed to prefix with `methods::`, so I have reverted to attaching `methods`, with a `packageStartupMessage` alerting the user that it is being attached.

**Comment by bfgray3** [comment]
@thirdwing I am now getting the following testing failure.  I don't believe I saw this before; has there been a change in the testing environment?

```
1. Error: UNET (@test_img_seg.R#95) --------------------------------------------
there is no package called 'imager'
1: library(imager) at R-package/tests/testthat/test_img_seg.R:95
2: stop(txt, domain = NA)
```

**Comment by thirdwing** [comment]
I am not sure about changes in the testing environment.

In your case, it seems `libxt-dev` hasn't been installed.

File: R-package/R/zzz.R [hunk 1]
```diff
@@ -16,7 +16,10 @@ NULL
 
 .onLoad <- function(libname, pkgname) {
   # Require methods for older versions of R
-  require(methods)
+  if (! is.element("package:methods", search())) {
```
Comment (thirdwing) [inline R-package/R/zzz.R]: We required R >= 3.3.0 and I think after R 3.3.0 `methods` will always be loaded during the startup.

Comment (bfgray3) [reply R-package/R/zzz.R]: Cool, I hadn't paid attention to what R version was required.. Hadley Wickham has a nice writeup about how to deal with this issue [here](http://r-pkgs.had.co.nz/namespace.html).  He says `Rscript` doesn't automatically load `methods` like an interactive session does, so the following is useful:

```
#' @import methods
NULL
```

**Comment by bfgray3** [comment]
I'm stumped on how to install `libxt-dev`. Doesn't the testing environment build everything from scratch for every commit, so in particular wouldn't this library get installed?  I have also noticed the following two errors, which I believe at least in part lead to `imager`'s inability to load:

```
libdc1394 error: Failed to initialize libdc1394
```
and
```
xlib-backend.c:34:74: fatal error: X11/Intrinsic.h: No such file or directory
```

**Comment by bfgray3** [comment]
I wonder if #7747 will help with my problems.

**Comment by thirdwing** [comment]
The simplest way to fix your error is to change this line (https://github.com/apache/incubator-mxnet/blob/master/tests/ci_build/install/ubuntu_install_r.sh#L27) into 

```
apt-get install -y r-base r-base-dev libxml2-dev libssl-dev libxt-dev
```

Just adding `libxt-dev` should fix the testing.

`libxt-dev` should be installed by default and this is a little weird.

**Comment by thirdwing** [comment]
Two more points: (1) the `libdc1394 error` can be ignored for now. It shouldn't cause any further problems; (2) personally, I want to clean up and merge #7476 before your PR. I appreciate your understanding.

**Comment by bfgray3** [comment]
Of course.  Thanks for your patience.

**Comment by bfgray3** [comment]
@thirdwing I have incorporated the updates from #7476 into my pull request and I now have a green check for the continuous integration tests.  Please let me know your thoughts at your convenience.  Thank you.

**Comment by piiswrong** [comment]
The tests are still failing for some reason.

Could you rebase your branch on current master and try again?

**Comment by bfgray3** [comment]
Sorry--I had though all tests were green--I think the jenkins CI might have run a little after the appveyor.  I am struggling on the rebase because of some conflicts.  I could make a new PR if that is alright with you, which would include a cleaner commit history.  Please let me know.

**Pull request closed without merging**